stacks-signer/src/stacks_client.rs

```diff
@@ -115,3 +356,88 @@ fn slot_id(id: u32, message: &Message) -> u32 {
     };
     SLOTS_PER_USER * id + slot_id
 }
+
+#[cfg(test)]
+mod tests {
+    use std::{
+        io::{Read, Write},
+        net::{SocketAddr, TcpListener},
+        thread::spawn,
+    };
+
+    use super::*;
+
+    struct TestConfig {
+        mock_server: TcpListener,
+        client: StacksClient,
+    }
+
+    impl TestConfig {
+        pub fn new() -> Self {
+            let mut config = Config::load_from_file("./src/tests/conf/signer-0.toml").unwrap();
+
+            let mut mock_server_addr = SocketAddr::from(([127, 0, 0, 1], 0));
+            // Ask the OS to assign a random port to listen on by passing 0
+            let mock_server = TcpListener::bind(mock_server_addr).unwrap();
+
+            // Update the config to use this port
+            mock_server_addr.set_port(mock_server.local_addr().unwrap().port());
+            config.node_host = mock_server_addr;
+
+            let client = StacksClient::from(&config);
+            Self {
+                mock_server,
+                client,
+            }
+        }
+    }
+
+    fn write_response(mock_server: TcpListener, bytes: &[u8]) -> [u8; 1024] {
+        debug!("Writing a response...");
+        let mut request_bytes = [0u8; 1024];
+        {
+            let mut stream = mock_server.accept().unwrap().0;
+            let _ = stream.read(&mut request_bytes).unwrap();
+            stream.write_all(bytes).unwrap();
+        }
+        request_bytes
+    }
+
+    #[test]
+    fn read_only_contract_call_should_succeed() {
+        let config = TestConfig::new();
+        let h = spawn(move || {
+            config.client.read_only_contract_call(
+                &config.client.stacks_address,
+                "contract-name",
+                "function-name",
+                &[],
+            )
+        });
+        write_response(
+            config.mock_server,
+            b"HTTP/1.1 200 OK\n\n{\"okay\":true,\"result\":\"0x070d0000000473425443\"}",
+        );
+        let result = h.join().unwrap().unwrap();
+        assert_eq!(result, "0x070d0000000473425443");
+    }
+
+    #[test]
+    fn read_only_contract_call_should_fail() {
+        let config = TestConfig::new();
+        let h = spawn(move || {
+            config.client.read_only_contract_call(
+                &config.client.stacks_address,
+                "contract-name",
+                "function-name",
+                &[],
+            )
+        });
+        write_response(
+            config.mock_server,
+            b"HTTP/1.1 200 OK\n\n{\"okay\":false,\"cause\":\"Some reason\"}",
+        );
```

Comment:
You might get a 400 or a 404 as well.  Can you test these conditions too?

Reply:
Done.